src/box/lua/trigger.c

```diff
@@ -0,0 +1,272 @@
+ */
+static int
+luaT_trigger_set(struct lua_State *L)
+{
+	const char *event_name = luaL_checkstring(L, 1);
+	const char *trigger_name = luaL_checkstring(L, 2);
+	if (lua_gettop(L) < 3 || !luaL_iscallable(L, 3))
+		luaL_typerror(L, 3, "callable");
+	/*
+	 * The following code is written in assumption no error will be thrown.
+	 */
+	struct event *event = event_registry_get(event_name, true);
+	assert(event != NULL);
+	struct func_adapter *func = func_adapter_lua_create(L, 3);
+	struct event_trigger *trg = event_trigger_create(func, trigger_name);
+	event_reset_trigger(event, trigger_name, trg, NULL);
+	func_adapter_lua_get_func(func, L);
+	return 1;
+}
+
+/**
+ * Deletes a trigger with passed name from passed event.
+ * The first argument is event name, the second one is trigger name.
+ * Returns deleted trigger handler.
+ */
+static int
+luaT_trigger_del(struct lua_State *L)
+{
+	const char *event_name = luaL_checkstring(L, 1);
+	struct event *event = event_registry_get(event_name, false);
+	if (event == NULL)
+		return 0;
+	const char *trigger_name = luaL_checkstring(L, 2);
+	struct event_trigger *old;
+	event_reset_trigger(event, trigger_name, NULL, &old);
+	if (old != NULL) {
+		func_adapter_lua_get_func(old->func, L);
+		event_trigger_unref(old);
+	} else {
+		lua_pushnil(L);
+	}
+
+	return 1;
+}
+
+/**
+ * A callback for event_foreach function that calls a trigger.
+ * Since the only supported language is Lua, all the handlers are Lua objects.
+ */
+static int
+trigger_call_cb(struct event_trigger *trigger, void *arg)
+{
+	struct lua_State *L = (struct lua_State *)arg;
+	int top_svp = lua_gettop(L);
+	func_adapter_lua_get_func(trigger->func, L);
+	/*
+	 * Starts from 2 because the first value in the stack is an event name.
+	 */
+	for (int i = 2; i <= top_svp; ++i) {
+		lua_pushvalue(L, i);
+	}
+	int rc = luaT_call(L, top_svp - 1, LUA_MULTRET);
+	lua_settop(L, top_svp);
+	return rc;
+}
+
+/**
+ * Calls all the triggers registered on passed event with variable number of
+ * arguments. Execution is stopped by a first exception.
+ * First argument must be a string - all the other arguments will be passed
+ * to the triggers without any processing or copying.
+ * Returns no values on success. If one of the triggers threw an error, it is
+ * raised again.
+ */
+static int
+luaT_trigger_call(struct lua_State *L)
+{
+	const char *event_name = luaL_checkstring(L, 1);
+	struct event *event = event_registry_get(event_name, false);
+	if (event == NULL)
+		return 0;
+	if (event_foreach(event, trigger_call_cb, L) != 0)
+		return luaT_error(L);
+	return 0;
+}
+
+/**
+ * An argument for callback that pushes triggers of one event onto Lua stack.
+ */
+struct push_trigger_arg {
+	/**
+	 * Current index in an array.
+	 */
+	size_t idx;
+	/**
+	 * Lua stack.
+	 */
+	lua_State *L;
+};
+
+/**
+ * Appends an array [trigger_name, trigger_handler] to a pre-created array.
+ */
+static int
+luaT_trigger_info_push_trigger(struct event_trigger *trigger, void *arg)
```

Comment:
Please remove luaT prefix.

Reply:
Done. Left it only for public functions and methods.